### PR TITLE
Add test to ensure that servers listen on IPv4 and IPv6 for default host

### PR DESCRIPTION
**Pull request by errorxyz**
#### Description

Add tests to check that we listen on both IPv4 and IPv6 (if supported) over TCP and UDP.
~Update `TunnelLayer` docstrings.~
refs https://github.com/mitmproxy/mitmproxy_rs/pull/167

#### Checklist

 - [x] I have updated tests where applicable.
 - [ ] I have added an entry to the CHANGELOG.

### CODE REVIEW

File: test/mitmproxy/proxy/test_mode_servers.py [hunk 1]
```diff
@@ -289,7 +291,25 @@ async def test_udp_start_error():
...
+    try:
+        with socket.socket(socket.AF_INET6, socket.SOCK_STREAM) as sock:
+            sock.bind(("::1", 0))
+        return True
+    except OSError as e:
+        if e.errno in _ADDR_NOT_AVAIL:
+            return False
+        if e.errno in _ADDR_IN_USE:
+            return True
+        raise e
```
Comment (mhils) [inline test/mitmproxy/proxy/test_mode_servers.py:309]: The simple logic from https://github.com/mitmproxy/mitmproxy/blob/aa73608ef60c56081dabf1c39e6f0d783e4a208f/test/conftest.py#L18-L27 did not work? 😅 

Can we really get _ADDR_IN_USE when binding to port 0? 🤔

File: test/mitmproxy/proxy/test_mode_servers.py [hunk 2]
```diff
@@ -289,7 +291,25 @@ async def test_udp_start_error():
...
+
+
+async def test_dual_stack(caplog_async):
```
Comment (mhils) [inline test/mitmproxy/proxy/test_mode_servers.py:312]:
```suggestion
async def test_dual_stack(caplog_async):
    """Test that a server bound to "" binds on both IPv4 and IPv6 for both TCP and UDP."""
```

**Review by mhils — COMMENTED** [review]
Looks good so far!

File: test/mitmproxy/proxy/test_mode_servers.py (hunk 1 above)
Comment (errorxyz) [reply test/mitmproxy/proxy/test_mode_servers.py:309]: > The simple logic from

Thanks for pointing this one out, didn't strike me to look out for existing code

> Can we really get _ADDR_IN_USE when binding to port 0? 🤔

Just in [case](https://stackoverflow.com/questions/33813701/on-linux-when-performing-a-socket-bind-with-port-0-pick-a-random-port-using-c)?😅 But yes, makes more sense to go with the existing decorator

File: test/conftest.py [hunk 3]
```diff
@@ -15,16 +15,19 @@
...
+        s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
+        s.bind(("::1", 0))
+        s.close()
+    except OSError:
+        return True
+    else:
+        return False
+
+
+skip_no_ipv6 = pytest.mark.skipif(no_ipv6(), reason="Host has no IPv6 support")
```
[the remark this answers is not in the document]
Comment (errorxyz) [reply test/conftest.py:30]: oops 😅

**Review by mhils — APPROVED** [review]

**Pull request merged**